### PR TITLE
[Traceable FSDP2] Add all_gather_into_tensor out variant

### PR DESCRIPTION
**Pull request by yf225**
This PR adds `torch.ops._c10d_functional.all_gather_into_tensor_out`.

It's important for tracing FSDP2, because FSDP2 pre-allocates the output buffer of AllGather, and makes input buffer an alias of the output buffer, and expects both of them to be used to achieve lower memory usage. If we don't preserve this behavior and instead functionalize the AllGather op, AllGather op will then create a brand-new output buffer (instead of reusing), thus significantly increasing the memory usage.

The expectation is that we will "re-inplace" the AllGather op by switching to the out variant in Inductor post-grad stage via an FX pass, so this API is not expected to be directly used by users.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k

### CODE REVIEW

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 1]
```diff
@@ -321,6 +334,13 @@ TORCH_LIBRARY(_c10d_functional, m) {
           c10::DispatchKey::CompositeExplicitAutograd, ::all_reduce_coalesced_),
       {at::Tag::pt2_compliant_tag});
 
+  m.def(
+      "all_gather_into_tensor_(Tensor(a!) output, Tensor input, int group_size, str group_name) -> Tensor(a!)",
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/Functional.cpp:338]: I would really love this to be an out-variant API, this is not a inplace allgather from operator prospective, it's a out variant op, we should follow the aten naming convention to make the op be an actual out-variant

Comment (yf225) [reply torch/csrc/distributed/c10d/Functional.cpp]: Sounds great! Updated.

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 2]
```diff
@@ -321,6 +334,13 @@ TORCH_LIBRARY(_c10d_functional, m) {
           c10::DispatchKey::CompositeExplicitAutograd, ::all_reduce_coalesced_),
       {at::Tag::pt2_compliant_tag});
 
+  m.def(
+      "all_gather_into_tensor_out(Tensor input, int group_size, str group_name, *, Tensor(a!) out) -> Tensor(a!)",
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/Functional.cpp:338]: Nice! We can establish the convention that:
- Collectives that modify the input are postfixed with `_`
- Out-variant of collectives are postfixed with `_out`, and the `out` argument is keyword-only

**Review by yifuwang — APPROVED** [review]
Looks great!

**Review by wanchaol — APPROVED** [review]
lgtm!

**Pull request closed without merging**